File: programmers/42586.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {

    vector<int> round, temp;
    vector<vector<int>> stack;
    vector<int> answer;

    int index(0),counter(0);

    for (int i = 0; i < progresses.size(); ++i) {
        round.push_back((100 - progresses[i]) / speeds[i]);
        if (((100-progresses[i]) % speeds[i]) != 0) round[i]++;
    }

    while (index < round.size()) {

        temp.push_back(round[index]);
        counter = round[index];
        if (++index < round.size()) {
            while(round[index] <= counter) {
                temp.push_back(round[index]);
                counter = counter - round[index];
                index++;
            }
        }
        stack.push_back(temp);
        temp.clear();
    }
    

    for (auto result : stack) answer.push_back(result.size());

    return answer;
}

int main() {
    vector<int> progress = {93, 30, 55};
    vector<int> speed = {1, 30, 5};

    for (auto i : solution(progress, speed)) cout<< i << endl;

    return 0;

}

File: programmers/42586.py
from collections import deque
import math

# no zip edition
def solution(progresses, speeds):
    round = deque([])
    answer = []

    for i in range(len(progresses)):
        round.append((100 - progresses[i]) // speeds[i])
        if ((100 - progresses[i]) % speeds[i] != 0):
            round[i] += 1
    
    while (len(round) > 0):
        temp = round.popleft()
        func = 1
        while ((len(round) > 0) and (round[0] <= temp)):
            round.popleft()
            func += 1
        answer.append(func)


    return answer

# zip editon
def solution_zip(progresses, speeds):
    answer = []
    temp = 0
    for p, s in zip(progresses, speeds):
        current = math.ceil((100-p)/s)
        if len(answer) == 0 or current > temp:
            answer.append(1)
            temp = current
        else:
            answer[-1] += 1
    
    return answer

